### PR TITLE
DEP: deprecate `product`, `cumproduct`, `sometrue`, `alltrue`

### LINKED ISSUE #14584: Status of aliased/wrapped functions, including `numpy.cumproduct`, `np.alltrue`, `np.sometrue`.
These functions exist and are part of the public API.

* `round_` (has docstring, has documentation)
* `product` (has docstring, no documentation)
* `cumproduct` (has docstring, no documentation)
* `sometrue` (has docstring, no documentation)
* `alltrue` (has docstring, no documentation)
* `rank` (has docstring, no documentation)

they also all trivially alias/wrap something else in the API that is more prevalent and clearly supported/documented.

Numba has just received PRs to enable support for a few of these functions, but I am now wondering if they are actually needed in NumPy itself. There seems to be limited internal references in the NumPy code base to the above.

This ticket is of a similar ilk to https://github.com/numpy/numpy/issues/14155.

Thanks in advance for your input!

### PR DESCRIPTION
**Pull request by rgommers**
Closes gh-14584

As a note for reviewers: these functions were already not present in the reference guide, so no changes were needed there.

### CODE REVIEW

File: numpy/core/fromnumeric.py [hunk 1]
```diff
@@ -3870,8 +3891,15 @@ def alltrue(*args, **kwargs):
...
 
+    .. deprecated:: 1.25.0
+        ``alltrue`` is deprecated as of NumPy 1.25.0, and will be
+        removed in NumPy 2.0. Please use `all` instead.
+
     See Also
     --------
     numpy.all : Equivalent function; see for details.
     """
+    warnings.warn("`alltrue` is deprecated as of NumPy 1.25.0, and will be "
```
[the remark this answers is not in the document]
Comment (rgommers) [reply numpy/core/fromnumeric.py:3902]: Good point, thanks. Fixed now.

**Comment by charris** [comment]
Thanks Ralf.

File: numpy/core/fromnumeric.py [hunk 2]
```diff
@@ -3832,10 +3833,18 @@ def product(*args, **kwargs):
...
+        removed in NumPy 2.0. Please use `prod` instead.
+
     See Also
     --------
     prod : equivalent function; see for details.
     """
+    # 2023-03-02, 1.25.0
+    warnings.warn("`product` is deprecated as of NumPy 1.25.0, and will be "
+                  "removed in NumPy 2.0. Please use `prod` instead.",
+                  DeprecationWarning, stacklevel=2)
```
Comment (seberg) [inline numpy/core/fromnumeric.py:3847]: I don't want to sweat it, but to note: it is slightly safer to put this into the dispatcher and (conditionally) here.  We do that in some/many/all of such deprecations.
The point being that a downstream user can be affected but not use the NumPy implementation.  (Downstream libs are likely to copy the warning, although if it was on the dispatcher they wouldn't have to.)

Comment (rgommers) [reply numpy/core/fromnumeric.py:3847]: Hmm, that is true technically. I can't find any example of this in the code base right now. Do you happen to remember an example? If we want to do this structurally, then maybe it's also useful to update the backwards compatibility NEP with a note/example.

Comment (seberg) [reply numpy/core/fromnumeric.py:3847]: There is half of an example right now, but the stacking functions used (and still use) the pattern with `overrides.ARRAY_FUNCTION_ENABLED`.

Dunno if we have any other deprecations in flight that should potentially be using the pattern...

They always call the dispatcher (or part of it) to check for sequence in put, that check used to be a warning.

Comment (rgommers) [reply numpy/core/fromnumeric.py:3847]: Followed up on this in gh-23364

**Pull request merged**